src/root.zig

```diff
@@ -114,55 +114,80 @@ pub fn AlgorithmType(
 
         allocator: std.mem.Allocator,
 
+        initialised: bool,
+
         pub fn deinit(self: *Self) void {
-            self.m.deinit();
-            self.x.deinit();
-            self.m_skip.deinit();
-            self.allocator.free(self.role_bonus);
-            self.allocator.free(self.bonus_buffer);
-            self.allocator.free(self.first_match_buffer);
-            self.allocator.free(self.traceback_buffer);
+            self.deallocateMatrixAndBuffer();
             self.* = undefined;
         }
 
         pub fn init(
             allocator: std.mem.Allocator,
+        ) !Self {
+            return .{
+                .m = undefined,
+                .x = undefined,
+                .m_skip = undefined,
+
+                .role_bonus = undefined,
+                .bonus_buffer = undefined,
+                .first_match_buffer = undefined,
+                .traceback_buffer = undefined,
+                .allocator = allocator,
+                .initialised = false,
+            };
```

Comment:
```suggestion
        haystack_size,
        needle_size,
        ) !Self {
            var self: Self = undefined;
            self.allocator = allocator;
            self.allocateMatrixBuffer(haystack_size, needle_size);
```
Let's not return uninitialized state (RAII) -- instead use the allocation function in the init. Can then drop the `initialised` boolean field.